[SPARK-17500][PySpark]Make DiskBytesSpilled metric in PySpark shuffle right

libratiger opened the pull request:
## What changes were proposed in this pull request?

The origin way increases the DiskBytesSpilled metric with the file size during each spill in ExternalMerger && ExternalGroupBy, but we only need the last size.
## How was this patch tested?

No extra tests, because this just update the metrics

Commit:
Make DiskBytesSpilled metric in PySpark shuffle right

The origin way increase the DiskBytesSpilled metric with the file
size during each spill in ExternalMerger && ExternalGroupBy, but we only need the last size.

No extra Tests, because this just update the metrics

Author: Li Hu <lihu@bytedance.com>

AmplabJenkins commented:
Can one of the admins verify this patch?

djvulee commented:
@srowen  @davies  mind taking a look? This PR is very simple.

srowen commented:
Given how DiskBytesSpilled is used, and still used in other parts of the code, this doesn't look correct. It seems to be a global that is always incremented. Here you reset the value in certain cases, effectively.

djvulee commented:
@srowen  you are right, I will correct it soon.

Commit:
Correct the DiskBytesSpilled metrics

We do not consider the DiskBytesSpilled can be increased by other
methods at the prev commit

djvulee commented:
@srowen  I update PR using an increment way to update the DiskBytesSpilled metrics.

srowen commented:
Is the idea that the file may be non empty when written ?
There is at least one more instance of this call but maybe the file is known to be empty before.

djvulee commented:
@srowen No. It does not matter whether the file is empty or not, if the file is empty, the `getsize()` just return 0, and this should be OK.

srowen commented:
I get that, but if it's always true, then there was no problem to begin with. That's what the code seems to think right now. I haven't looked at the code much but that's the question -- are you sure the files are non-empty in some scenario?

djvulee commented:
@srowen Yes, the file seems always empty before write, so the origin way is OK.  Sorry for this PR is not thoughtful enough, I just get a mislead by the other method in the shuffle.py, which used the pos before write. 

I will close this PR.

srowen commented:
OK well if you see evidence later that the disk spilled bytes are unreasonably high, it's worth reinvestigating to see if there's a problem like this. If you aren't seeing bad metrics though, then maybe there's no problem.
